R/collect.R

```diff
@@ -373,10 +373,22 @@ collector <- function(x, coll_col = ".predictions") {
   param_names <- .get_tune_parameter_names(x)
   tibble_metrics <- purrr::map_lgl(x[[".metrics"]], tibble::is_tibble)
   x <- x[tibble_metrics, ]
-  x <- dplyr::select(x, dplyr::any_of(".iter"), .metrics)
-  x <- tidyr::unnest(x, cols = .metrics)
-  x <- dplyr::select(x, dplyr::all_of(param_names), ".config", dplyr::any_of(".iter"))
-  dplyr::distinct(x)
+  x <- x[, colnames(x) %in% c(".iter", ".metrics")]
+
+  res <- vctrs::list_unchop(x$.metrics)
+  if (".iter" %in% colnames(x)) {
+    res <- vctrs::vec_cbind(
+      res,
+      vctrs::data_frame(
+        .iter = vctrs::vec_rep_each(x$.iter, vctrs::list_sizes(x$.metrics))
+      )
+    )
+  }
+
+  keep_cols <- colnames(res) %in% c(param_names, ".config", ".iter")
+  res <- res[, keep_cols]
+
+  dplyr::distinct(res)
```

Comment:
There is nothing functionally wrong with this at all, it is a nice improvement to switch to vctrs.

In the spirit of comparing against what I would have written, here is my attempt. Feel free to ignore it if you like yours:

```r
metrics <- x[[".metrics"]]

out <- vctrs::list_unchop(metrics)
out <- out[c(param_names, ".config")]

if (has_name(x, ".iter")) {
  iter <- x[[".iter"]]
  out[[".iter"]] <- vctrs::vec_rep_each(iter, times = vctrs::list_sizes(metrics))
}

out <- vctrs::vec_unique(out)

out
```

Reply:
Awesome, thank you! Applied in [2995500](https://github.com/tidymodels/tune/pull/656/commits/29955001c43c674efafefadf071bda9151d4d843).